plugins/auth-backend/src/providers/cloudflare-access/provider.ts

```diff
@@ -0,0 +1,345 @@
+  private readonly jwtKeySet: any;
+  private readonly logger: Logger;
+  private readonly cache?: CacheClient;
+
+  constructor(options: Options) {
+    this.teamName = options.teamName;
+    this.authHandler = options.authHandler;
+    this.signInResolver = options.signInResolver;
+    this.resolverContext = options.resolverContext;
+    this.jwtKeySet = createRemoteJWKSet(
+      new URL(
+        `https://${this.teamName}.cloudflareaccess.com/cdn-cgi/access/certs`,
+      ),
+    );
+    this.logger = options.logger;
+    this.cache = options.cache;
+  }
+
+  frameHandler(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  async refresh(req: express.Request, res: express.Response): Promise<void> {
+    // ProxiedSignInPage calls `/refresh` implicitly each time the backstage
+    // app is refreshed on the browser.
+    // User authentication is then checked here.
+    const result = await this.getResult(req);
+    const response = await this.handleResult(result);
+    res.json(response);
+  }
+
+  start(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  private async getIdentityProfile(
+    jwt: string,
+  ): Promise<CloudflareAccessIdentityProfile> {
+    const headers = new Headers();
+    // set both headers just the way inbound responses are set
+    headers.set(CF_JWT_HEADER, jwt);
+    headers.set('cookie', `${COOKIE_AUTH_NAME}=${jwt}`);
+    try {
+      const res: FetchResponse = await fetch(
+        `https://${this.teamName}.cloudflareaccess.com/cdn-cgi/access/get-identity`,
+        { headers },
+      );
+      if (!res.ok) {
+        // Cast to work around https://github.com/backstage/backstage/issues/12166
+        throw await ResponseError.fromResponse(res as unknown as Response);
+      }
+      const cfIdentity = await res.json();
+      return cfIdentity as unknown as CloudflareAccessIdentityProfile;
+    } catch (err) {
+      this.logger.error(`getIdentityProfile failed: ${err}`);
+      return Promise.reject(err);
+    }
+  }
+
+  private async getResult(
+    req: express.Request,
+  ): Promise<CloudflareAccessResult> {
+    // JWTs generated by Access are available in a request header as
+    // Cf-Access-Jwt-Assertion and as cookies as CF_Authorization.
+    let jwt = req.header(CF_JWT_HEADER);
+    if (!jwt) {
+      jwt = req.cookies.CF_Authorization;
+    }
+    if (!jwt) {
+      // Only throw if both are not provided by Cloudflare Access since either
+      // can be used.
+      throw new AuthenticationError(
+        `Missing ${CF_JWT_HEADER} from Cloudflare Access`,
+      );
+    }
+
+    // Cloudflare signs the JWT using the RSA Signature with SHA-256 (RS256).
+    // RS256 follows an asymmetric algorithm; a private key signs the JWTs and
+    // a separate public key verifies the signature.
+    const verifyResult = await jwtVerify(jwt, this.jwtKeySet, {
+      issuer: `https://${this.teamName}.cloudflareaccess.com`,
+    });
+    const cfAccessResultStr = await this.cache?.get(jwt);
```

Comment:
Perhaps use `verifyResult.payload.sub` as the key here instead? Or something like it. Just to remove the risk of leaking JWTs through the cache. Could also hash it too, so something like `sha1(jwt)` as the key.